drivers/CpuUid.cpp

```diff
@@ -0,0 +1,70 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "hal/cpu_uid_api.h"
+#include "platform/mbed_assert.h"
+#include "platform/CriticalSectionLock.h"
+#include "drivers/CpuUid.h"
+
+#if DEVICE_CPUUID
+
+namespace mbed {
+
+uint8_t CpuUid::_uidbuf[MBED_CPU_UID_SIZE] = {0};
+uint8_t* CpuUid::_uidptr = NULL;
+char CpuUid::_strbuf[CPU_UID_STRING_BUFFER_SIZE] = {'\0'};
+char* CpuUid::_strptr = NULL;
+#ifndef MBED_CPU_UID_STR_SIZE_MAX
+const char CpuUid::_hexChars[16] = {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F'};
+#endif
+
+CpuUid::CpuUid()
+{
+    populate_uid_buf();
+}
+
+void CpuUid::populate_uid_buf()
+{
+    if (_uidptr == NULL) {
+        CriticalSectionLock lock;
```

Comment:
What is the reason for have critical section hera? nRF52 doesn't requires a such. Does any other target need this?

Reply:
imagine the scenario where the uid buffer is not populated yet ...

now there could be one part of code somewhere which uses CpuUid the first time - the populate_uid_buf() function will be called ... if there is another part of code somewhere in an interrupt function, which also uses a CpuUid object, and that interrupt fires exactly in the middle of the populate_uid_buf() function, then there could be a messup theoretically.
Maybe it is not needed, because the only thing that is done here is setting the pointer to the static buffer (which will always have the same address) ... but i wanted to be prepared if some more code will be added here

Reply:
I see posybilisty to simply move initialization of this UUID to early startup instead that. Thanks to that critical lock won't be necessary anymore. This just another option.

Reply:
yeah that would be a possibility.

but it seems there is one philosophical question here: to use only local buffers (reading from hal each time the driver is read), or do some "caching" in driver's members.

I do not know how to unite those two paradigms

Reply:
That's right - Its you call to decide.